File: creational/factory_method/ruby/lib/creator/notification_sender.rb
require_relative '../product/notification'

class NotificationSender
  def send_notification(recipient, message)
    notification = create_notification
    puts "Sending notification: #{notification.class}"
    notification.send(recipient, message)
    puts "Notification sent successfully"
  end

  def create_notification()
    raise NotImplementedError, "Subclasses must implement this method"
  end
end

File: creational/factory_method/java/src/creater/NotificationSender.java
package creater;

import product.Notification;

public abstract class NotificationSender {
    private Notification notification;

    public void sendAlert(String message) {
      this.notification = createNotification();

      System.out.println("Sending notification from the NotificationSender");
      notification.send(message);
      System.out.println("Notification sent from the NotificationSender");
    }

    public abstract Notification createNotification();
}
